proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,489 @@
+
+## How to try KEEEP-87? (First approach)
+
+- Clone [Our Kotlin fork](https://github.com/arrow-kt/kotlin) and checkout the **keep-87** branch.
+- Follow the instructions on the [README](https://github.com/arrow-kt/kotlin/blob/master/ReadMe.md#build-environment-requirements) configure the necessary JVMs.
+- Follow the instructions on the [README](https://github.com/arrow-kt/kotlin/blob/master/ReadMe.md#-working-with-the-project-in-intellij-idea) to open the project in IntelliJ IDEA.
+- Once you have everything working, you can run a new instance of IntelliJ IDEA with the new modifications to the language by executing `./gradlew runIde`. There is also a pre-configured run configuration titled **IDEA** that does this.
+- It will open a new instance of the IDE where you can create a new project and experiment with the new features of the language. You can also download [this project](https://github.com/arrow-kt/kotlin/files/3064100/Keep87Sample.zip)
+) with some sample code that you can try out.
+
+## How to try KEEEP-87? (Alternative approach - easier)
+
+We've got the Keep 87 deployed to our own Idea plugin repository over Amazon s3. To use it:
+
+- Download the latest version of IntelliJ IDEA 2018.2.4 from JetBrains
+- Go to `preferences` -> `plugins` section.
+- Click on "Manage Plugin Repositories".
+![InstallKeepFromRepository1](https://user-images.githubusercontent.com/6547526/55884351-38609d80-5ba8-11e9-8855-3c570ee8a1af.png)
+- Add our Amazon s3 plugin repository as in the image.
+![InstallKeepFromRepository2](https://user-images.githubusercontent.com/6547526/55884427-562e0280-5ba8-11e9-98e8-8811e8e3e8b0.png)
+- Now browse for "keep87" plugin.
+![InstallKeepFromRepository3](https://user-images.githubusercontent.com/6547526/55884468-67770f00-5ba8-11e9-92d6-e9cc8cc3f572.png)
+- Install it.
+![InstallKeepFromRepository4](https://user-images.githubusercontent.com/6547526/55884479-6b0a9600-5ba8-11e9-8a19-0eec53187fc5.png)
+- Download and run [this project](https://github.com/arrow-kt/kotlin/files/3064100/Keep87Sample.zip) on that IntellIJ instance.
+
+
+## What's still to be done?
+
+### Type-side implementations
+
+We have additional complications when you consider multi-parameter constraint interfaces.
+
+```kotlin
+package foo.repo
+
+// I stands for the index type, A for the stored type.
+interface Repository<I, A> {
+   ...
+}
+```
+
+```kotlin
+package data.foo
+
+data class Id(...)
+extension class RepoIndexedById<A> : Repository<Id, A> {
+   ...
+}
+```
+
+```kotlin
+package data.foo.user
+
+data class User(...)
+extension class UserRepository<I> : Repository<I, User> {
+   ...
+}
+```
+
+The above instances are each defined alongside their respective type definitions and yet they clearly conflict with each other. We will also run into quandaries once we consider generic types. We can crib some prior art from Rust<sup>1</sup> to help us out here.
```

Comment:
quandaries?

Reply:
"a state of perplexity or uncertainty over what to do in a difficult situation."